src/test/java/com/znsio/teswiz/runner/AppPathTest.java

```diff
@@ -0,0 +1,105 @@
+
+public class AppPathTest {
+    private static final String tempFolderPath = System.getProperty("user.dir") + File.separator + "temp";
+    private static final String expectedDirectoryPath = tempFolderPath + File.separator + "sampleApps";
+    private static final String fileName = "VodQA.apk";
+    private static final String expectedAppPath = expectedDirectoryPath + File.separator + fileName;
+    private static final String appPathAsUrl = "https://github.com/anandbagmar/sampleAppsForNativeMobileAutomation/raw/main/VodQA.apk";
+    private static final String appPathAsIncorrectUrl = "https://github.com/anandbagmar/sampleAppsForNativeMobileAutomation/ra/main/VodQA.apk";
+    private static final String appPathAsFilePath = expectedAppPath;
+    private static final String appPathAsIncorrectFilePath = tempFolderPath + File.separator + "smleApps" + File.separator + fileName;
+
+    @Test
+    void validateAppPathAsCorrectUrlCreateRepoAndDownloadFileWhenBothDoNotExist() {
+        deleteDirectorySubDirectoryAndFiles(tempFolderPath);
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsCorrectUrlDoNotDownloadFileWhenRepoAndFileAlreadyExist() {
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsCorrectUrlOnlyDownloadFileWhenRepoExist() {
+        deleteFile(expectedAppPath);
+        createDirectory(expectedDirectoryPath);
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsIncorrectUrlGivesError() {
+        deleteDirectorySubDirectoryAndFiles(tempFolderPath);
+        assertThrows(RuntimeException.class, () -> DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsIncorrectUrl));
+    }
+
+    @Test
+    void validateAppPathAsCorrectFilePathLoadsFileWhenAlreadyExist() {
+        DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        String actualAppPath = DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsFilePath);
+        assertEquals(expectedAppPath, actualAppPath);
+        assertTrue(Files.exists(Paths.get(actualAppPath)));
+    }
+
+    @Test
+    void validateAppPathAsIncorrectFilePathGivesError() {
+        DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsUrl);
+        assertThrows(RuntimeException.class, () -> DeviceSetup.convertAppPathToFilePathIfNeeded(appPathAsIncorrectFilePath));
+    }
+
+    @Test
+    void validateAppPathAsFilePathButFileNotPresentGivesError() {
```

Comment:
It would be better if we could specify the error thrown in the test name

Reply:
done